runtime/runtime-params-estimator/src/action_costs.rs

```diff
@@ -0,0 +1,626 @@
+        .verify_cost(&mut ctx.testbed())
+}
+
+pub(crate) fn stake_send_not_sir(ctx: &mut EstimatorContext) -> GasCost {
+    ActionEstimation::new(ctx)
+        .add_action(stake_action())
+        .predecessor(AccountRequirement::SameAsSigner)
+        .receiver(AccountRequirement::RandomUnused)
+        .verify_cost(&mut ctx.testbed())
+}
+
+pub(crate) fn stake_exec(ctx: &mut EstimatorContext) -> GasCost {
+    ActionEstimation::new(ctx)
+        .add_action(stake_action())
+        .predecessor(AccountRequirement::SameAsSigner)
+        .receiver(AccountRequirement::SameAsSigner) // staking must be local
+        .apply_cost(&mut ctx.testbed())
+}
+
+pub(crate) fn add_full_access_key_send_sir(ctx: &mut EstimatorContext) -> GasCost {
+    ActionEstimation::new(ctx)
+        .add_action(add_full_access_key_action())
+        .predecessor(AccountRequirement::SameAsSigner)
+        .receiver(AccountRequirement::SameAsSigner)
+        .verify_cost(&mut ctx.testbed())
+}
+
+pub(crate) fn add_full_access_key_send_not_sir(ctx: &mut EstimatorContext) -> GasCost {
+    ActionEstimation::new(ctx)
+        .add_action(add_full_access_key_action())
+        .verify_cost(&mut ctx.testbed())
+}
+
+pub(crate) fn add_full_access_key_exec(ctx: &mut EstimatorContext) -> GasCost {
+    ActionEstimation::new(ctx)
+        .add_action(add_full_access_key_action())
+        .inner_iters(1) // adding the same key a second time would fail
+        .predecessor(AccountRequirement::SameAsSigner)
+        .receiver(AccountRequirement::SameAsSigner)
+        .apply_cost(&mut ctx.testbed())
+}
+
+pub(crate) fn add_function_call_key_base_send_sir(ctx: &mut EstimatorContext) -> GasCost {
+    ActionEstimation::new(ctx)
+        .add_action(add_fn_access_key_action(ActionSize::Min))
+        .predecessor(AccountRequirement::SameAsSigner)
+        .receiver(AccountRequirement::SameAsSigner)
+        .verify_cost(&mut ctx.testbed())
+}
+
+pub(crate) fn add_function_call_key_base_send_not_sir(ctx: &mut EstimatorContext) -> GasCost {
+    ActionEstimation::new(ctx)
+        .add_action(add_fn_access_key_action(ActionSize::Min))
+        .verify_cost(&mut ctx.testbed())
+}
+
+pub(crate) fn add_function_call_key_base_exec(ctx: &mut EstimatorContext) -> GasCost {
+    ActionEstimation::new(ctx)
+        .add_action(add_fn_access_key_action(ActionSize::Min))
+        .inner_iters(1) // adding the same key a second time would fail
+        .predecessor(AccountRequirement::SameAsSigner)
+        .receiver(AccountRequirement::SameAsSigner)
+        .apply_cost(&mut ctx.testbed())
+}
+
+pub(crate) fn add_function_call_key_byte_send_sir(ctx: &mut EstimatorContext) -> GasCost {
+    ActionEstimation::new(ctx)
+        .add_action(add_fn_access_key_action(ActionSize::Max))
+        .predecessor(AccountRequirement::SameAsSigner)
+        .receiver(AccountRequirement::SameAsSigner)
+        .verify_cost(&mut ctx.testbed())
+        / ActionSize::Max.key_methods_list()
```

Comment:
Observation: `add_fn_acceess_key_action` generates 4 bytes of method names `key_methods_list` times so isn’t this actually  a cost/32-bit-word not cost/byte?

Reply:
I have to admit it's not pretty, but the factor 4 is calculated in here:
```rust
fn add_fn_access_key_action(size: ActionSize) -> Action {
    // 3 bytes for "foo" and one for an implicit separator
    let method_names = vec!["foo".to_owned(); size.key_methods_list() as usize / 4];
```

Note the hidden `/ 4`. Therefore the final size generated is exactly `ActionSize::Max.key_methods_list()` and if my Math is right, it ends up as cost per byte.

Reply:
Ah, hm. This is going to floor the numbers, but makes sense, I had missed this detail when looking through this code.